FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Multitask/DemoTasks/mqtt-agent-task.c

```diff
@@ -0,0 +1,955 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://github.com/FreeRTOS
+ *
+ */
+
+/*
+ * This demo creates multiple tasks, all of which use the MQTT agent API to
+ * communicate with an MQTT broker through the same MQTT connection.
+ *
+ * This file contains the initial task created after the TCP/IP stack connects
+ * to the network.  The task:
+ *
+ * 1) Connects to the MQTT broker.
+ * 2) Creates the other demo tasks, in accordance with the #defines set in
+ *    demo_config.h.  For example, if demo_config.h contains the following
+ *    settings:
+ *
+ *    #define democonfigCREATE_LARGE_MESSAGE_SUB_PUB_TASK     1
+ *    #define democonfigNUM_SIMPLE_SUB_PUB_TASKS_TO_CREATE 3
+ *
+ *    then the initial task will create the task implemented in
+ *    LargeMessageSubScribePublish.c and three instances of the task
+ *    implemented in SimpleSubscribePublishT.c.  See the comments at the top
+ *    of those files for more information.
+ *
+ * 3) After creating the demo tasks the initial task could create the MQTT
+ *    agent task.  However, as it has no other operations to perform, rather
+ *    than create the MQTT agent as a separate task the initial task just calls
+ *    the agent's implementing function - effectively turning itself into the
+ *    MQTT agent.
+ */
+
+
+/* Standard includes. */
+#include <string.h>
+#include <stdio.h>
+#include <assert.h>
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "queue.h"
+#include "task.h"
+
+/* FreeRTOS+TCP includes. */
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_Sockets.h"
+
+/* Demo Specific configs. */
+#include "demo_config.h"
+
+/* MQTT library includes. */
+#include "core_mqtt.h"
+
+/* MQTT agent include. */
+#include "mqtt_agent.h"
+
+/* MQTT Agent ports. */
+#include "freertos_agent_message.h"
+#include "freertos_command_pool.h"
+
+/* Exponential backoff retry include. */
+#include "backoff_algorithm.h"
+
+/* Subscription manager header include. */
+#include "subscription_manager.h"
+
+
+/* Transport interface include. */
+#if defined( democonfigUSE_TLS ) && ( democonfigUSE_TLS == 1 )
+    #include "using_mbedtls.h"
+#else
+    #include "using_plaintext.h"
+#endif
+
+/* This demo uses compile time options to select the demo tasks to created.
+ * Ensure the compile time options are defined.  These should be defined in
+ * demo_config.h. */
+#ifndef democonfigNUM_SIMPLE_SUB_PUB_TASKS_TO_CREATE
```

Comment:
What is the purpose of allowing a zero value for the config? Do we want to allow the demo to be run with just the Agent task and no MQTT operations..?

Reply:
No, that was copied over from the labs repo. I removed that statement